runtime-modules/referendum/src/lib.rs

```diff
@@ -0,0 +1,740 @@
+pub struct SealedVote<Hash, CurrencyBalance> {
+    commitment: Hash,
+    stake: CurrencyBalance,
+}
+
+#[derive(Clone, Encode, Decode, PartialEq, Eq, Debug)]
+pub enum ReferendumResult<ReferendumOption, VotePower> {
+    Winners(Vec<(ReferendumOption, VotePower)>),
+    ExtraWinners(Vec<(ReferendumOption, VotePower)>),
+    NotEnoughWinners(Vec<(ReferendumOption, VotePower)>),
+    NoVotesRevealed,
+}
+
+impl<T, U> Default for ReferendumResult<T, U> {
+    fn default() -> ReferendumResult<T, U> {
+        ReferendumResult::NoVotesRevealed
+    }
+}
+
+/////////////////// Trait, Storage, Errors, and Events /////////////////////////
+
+pub trait Trait<I: Instance>: system::Trait {
+    /// The overarching event type.
+    type Event: From<Event<Self, I>> + Into<<Self as system::Trait>::Event>;
+
+    /// Maximum number of options in one referendum.
+    type MaxReferendumOptions: Get<u64>;
+    type ReferendumOption: Parameter
+        + Member
+        + BaseArithmetic
+        + Codec
+        + Default
+        + Copy
+        + MaybeSerialize
+        + PartialEq
+        + From<u64>
+        + Into<u64>;
+
+    /// Currency balance used for stakes.
+    type CurrencyBalance: Parameter
+        + Member
+        + BaseArithmetic
+        + Codec
+        + Default
+        + Copy
+        + MaybeSerialize
+        + PartialEq
+        + From<u64>
+        + Into<u64>;
+
+    type VotePower: Parameter
+        + Member
+        + BaseArithmetic
+        + Codec
+        + Default
+        + Copy
+        + MaybeSerialize
+        + PartialEq
+        + From<u64>
+        + Into<u64>;
+
+    type VoteStageDuration: Get<Self::BlockNumber>;
+    type RevealStageDuration: Get<Self::BlockNumber>;
+
+    type MinimumStake: Get<Self::CurrencyBalance>;
+
+    fn is_super_user(account_id: &<Self as system::Trait>::AccountId) -> bool;
+
+    fn caclulate_vote_power(
+        account_id: &<Self as system::Trait>::AccountId,
+        stake: <Self as Trait<I>>::CurrencyBalance,
+    ) -> <Self as Trait<I>>::VotePower;
+
+    fn has_sufficient_balance(
+        account: &<Self as system::Trait>::AccountId,
+        balance: &Self::CurrencyBalance,
+    ) -> bool;
+    fn lock_currency(
+        account: &<Self as system::Trait>::AccountId,
+        balance: &Self::CurrencyBalance,
+    ) -> bool;
+    fn free_currency(
+        account: &<Self as system::Trait>::AccountId,
+        balance: &Self::CurrencyBalance,
+    ) -> bool;
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait<I>, I: Instance> as Referendum {
+        /// Current referendum stage
+        pub Stage get(fn stage) config(): (ReferendumStage, T::BlockNumber);
+
+        /// Options of current referendum
+        pub ReferendumOptions get(fn referendum_options) config(): Option<Vec<T::ReferendumOption>>;
+
+        /// Votes in current referendum
+        pub Votes get(fn votes) config(): map hasher(blake2_128_concat) T::AccountId => SealedVote<T::Hash, T::CurrencyBalance>;
```

Comment:
Votes should not be identified with an account, you vote as a member not as a raw account holder.

Reply:
Misunderstood this one. Will introduce members instead of accounts to all relevant places.

Reply:
Done